### PR TITLE
Add end navigation functionality to navigation view

### PR DESCRIPTION
**Pull request by Guardiola31337**
- Fixes https://github.com/mapbox/mapbox-navigation-android/issues/813

TODO

- [x] Add example to the test app
- [x] `MapboxMap.OnScrollListener` is deprecated, use `OnMoveListener` instead 👀 https://github.com/mapbox/mapbox-navigation-android/blob/cbbe8e0967927cccd4e567aa15b27194e3f9ce89/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationView.java#L91

### CODE REVIEW

**Comment by danesfeder** [comment]
@Guardiola31337 is this still WIP?  Just need to address the comments?

**Comment by Guardiola31337** [comment]
@danesfeder
> is this still WIP? Just need to address the comments?

I've addressed your comments in the last commit. In any case, this functionality touches a lot and should be thoroughly tested so I'd say yes still WIP so please I encourage you @danesfeder and @devotaaabel to pull the latest changes from `pg-end-navigation` and test all the scenarios that you can think of.

For example, I'm 👀 that `NavigationView#stopNavigation` doesn't crash even if you call it before `NavigationView#startNavigation`.

**Comment by Guardiola31337** [comment]
Also noticed https://github.com/mapbox/mapbox-navigation-android/issues/800 and run into the following :boom: 
```
06-05 19:19:29.878 24492-24492/com.mapbox.services.android.navigation.testapp E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.mapbox.services.android.navigation.testapp, PID: 24492
    java.lang.IllegalStateException: The specified child already has a parent. You must call removeView() on the child's parent first.
        at android.view.ViewGroup.addViewInner(ViewGroup.java:3882)
        at android.view.ViewGroup.addView(ViewGroup.java:3735)
        at android.view.ViewGroup.addView(ViewGroup.java:3680)
        at android.support.v7.widget.RecyclerView$5.addView(RecyclerView.java:778)
        at android.support.v7.widget.ChildHelper.addView(ChildHelper.java:107)
        at android.support.v7.widget.RecyclerView$LayoutManager.addViewInt(RecyclerView.java:8222)
        at android.support.v7.widget.RecyclerView$LayoutManager.addView(RecyclerView.java:8180)
        at android.support.v7.widget.RecyclerView$LayoutManager.addView(RecyclerView.java:8168)
        at android.support.v7.widget.LinearLayoutManager.layoutChunk(LinearLayoutManager.java:1573)
        at android.support.v7.widget.LinearLayoutManager.fill(LinearLayoutManager.java:1519)
        at android.support.v7.widget.LinearLayoutManager.onLayoutChildren(LinearLayoutManager.java:614)
        at android.support.v7.widget.RecyclerView.dispatchLayoutStep2(RecyclerView.java:3812)
        at android.support.v7.widget.RecyclerView.dispatchLayout(RecyclerView.java:3529)
        at android.support.v7.widget.RecyclerView.onLayout(RecyclerView.java:4082)
        at android.view.View.layout(View.java:15606)
        at android.view.ViewGroup.layout(ViewGroup.java:4968)
        at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1703)
        at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1557)
        at android.widget.LinearLayout.onLayout(LinearLayout.java:1466)
        at android.view.View.layout(View.java:15606)
        at android.view.ViewGroup.layout(ViewGroup.java:4968)
        at android.widget.RelativeLayout.onLayout(RelativeLayout.java:1076)
        at android.view.View.layout(View.java:15606)
        at android.view.ViewGroup.layout(ViewGroup.java:4968)
        at android.widget.RelativeLayout.onLayout(RelativeLayout.java:1076)
        at android.view.View.layout(View.java:15606)
        at android.view.ViewGroup.layout(ViewGroup.java:4968)
        at android.support.design.widget.CoordinatorLayout.layoutChild(CoordinatorLayout.java:1191)
        at android.support.design.widget.CoordinatorLayout.onLayoutChild(CoordinatorLayout.java:876)
        at android.support.design.widget.CoordinatorLayout.onLayout(CoordinatorLayout.java:895)
        at android.view.View.layout(View.java:15606)
        at android.view.ViewGroup.layout(ViewGroup.java:4968)
        at android.support.design.widget.CoordinatorLayout.layoutChild(CoordinatorLayout.java:1191)
        at android.support.design.widget.CoordinatorLayout.onLayoutChild(CoordinatorLayout.java:876)
        at android.support.design.widget.CoordinatorLayout.onLayout(CoordinatorLayout.java:895)
        at android.view.View.layout(View.java:15606)
        at android.view.ViewGroup.layout(ViewGroup.java:4968)
        at android.support.constraint.ConstraintLayout.onLayout(ConstraintLayout.java:1855)
        at android.view.View.layout(View.java:15606)
        at android.view.ViewGroup.layout(ViewGroup.java:4968)
        at android.widget.FrameLayout.layoutChildren(FrameLayout.java:573)
        at android.widget.FrameLayout.onLayout(FrameLayout.java:508)
        at android.view.View.layout(View.java:15606)
        at android.view.ViewGroup.layout(ViewGroup.java:4968)
        at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1703)
        at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1557)
        at android.widget.LinearLayout.onLayout(LinearLayout.java:1466)
        at android.view.View.layout(View.java:15606)
        at android.view.ViewGroup.layout(ViewGroup.java:4968)
        at android.widget.FrameLayout.layoutChildren(FrameLayout.java:573)
        at android.widget.FrameLayout.onLayout(FrameLayout.java:508)
        at android.view.View.layout(View.java:15606)
        at android.view.ViewGroup.layout(ViewGroup.java:4968)
        at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1703)
        at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1557)
        at android.widget.LinearLayout.onLayout(LinearLayout.java:1466)
        at android.view.View.layout(View.java:15606)
    	at android.view.
```
Might be related to #800

**Comment by Guardiola31337** [comment]
I'm running into https://github.com/mapbox/mapbox-navigation-android/issues/981 when rotating the screen 👀 

navigation_telemetry_leak

**Comment by Guardiola31337** [comment]
If you rotate the screen the screen multiple time along the route I'm getting the following 💥 

```
    --------- beginning of crash
06-11 20:52:48.987 8934-8934/com.mapbox.services.android.navigation.testapp E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.mapbox.services.android.navigation.testapp, PID: 8934
    java.lang.IndexOutOfBoundsException: Invalid index 0, size is 0
        at java.util.ArrayList.throwIndexOutOfBoundsException(ArrayList.java:255)
        at java.util.ArrayList.get(ArrayList.java:308)
        at com.mapbox.services.android.navigation.ui.v5.route.NavigationMapRoute$1.onProgressChange(NavigationMapRoute.java:177)
        at com.mapbox.services.android.navigation.v5.navigation.NavigationEventDispatcher.onProgressChange(NavigationEventDispatcher.java:145)
        at com.mapbox.services.android.navigation.v5.navigation.NavigationService.onNewRouteProgress(NavigationService.java:113)
        at com.mapbox.services.android.navigation.v5.navigation.NavigationEngine$1.run(NavigationEngine.java:92)
        at android.os.Handler.handleCallback(Handler.java:739)
        at android.os.Handler.dispatchMessage(Handler.java:95)
        at android.os.Looper.loop(Looper.java:148)
        at android.app.ActivityThread.main(ActivityThread.java:5417)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
```

rotation_multiple_time_on_progress_change_crash

I've also noticed that sometimes a instruction is played twice (same instruction almost at the same time) when you rotate the screen multiple times.

**Comment by Guardiola31337** [comment]
@devotaaabel 
> Found a few bugs which were mentioned on slack.

Are any of them not ticketed/reported here? If so, could you add them as comments please?

**Comment by devotaaabel** [comment]
Can you move the Paella activity below the NavigationViewActivity? Also, waypoint markers are still dropping off occasionally. Also, when cancelling, the route disappears. Not sure if this is by design or not.

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationView.java [hunk 1]
```diff
@@ -253,7 +247,13 @@ public boolean isRecenterButtonVisible() {
 
   @Override
   public void drawRoute(DirectionsRoute directionsRoute) {
-    navigationMap.drawRoute(directionsRoute);
+    if (navigationMap != null) {
```
Comment (danesfeder) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationView.java]: @Guardiola31337 let's double check that all `public` `NavigationView` APIs that use the `NavigationMapboxMap` are protected like this one

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationView.java [hunk 2]
```diff
@@ -323,21 +323,25 @@ public void updateCameraRouteOverview() {
    *
    * @param options with containing route / coordinate data
    */
+  @SuppressLint("MissingPermission")
```
Comment (danesfeder) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationView.java]: Is this annotation still necessary?

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationView.java [hunk 3]
```diff
@@ -323,21 +323,25 @@ public void updateCameraRouteOverview() {
...
     } else {
-      navigationMap.clearMarkers();
       navigationViewModel.updateNavigation(options);
     }
   }
 
+  /**
+   * Call this when the navigation session needs to end navigation without finishing the whole view
+   *
+   * @since 0.15.0
```
Comment (danesfeder) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationView.java]: Update javadoc `0.16.0`

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationView.java [hunk 4]
```diff
@@ -420,9 +424,11 @@ private void updateSavedInstanceStateMapStyle(@Nullable Bundle savedInstanceStat
    * @param bottomSheetState retrieved from savedInstanceState
    */
   private void resetBottomSheetState(int bottomSheetState) {
-    boolean isShowing = bottomSheetState == BottomSheetBehavior.STATE_EXPANDED;
-    summaryBehavior.setHideable(!isShowing);
-    summaryBehavior.setState(bottomSheetState);
+    if (bottomSheetState > 0) {
```
Comment (danesfeder) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationView.java]: Magic number

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/OnNavigationReadyCallback.java [hunk 5]
```diff
@@ -2,5 +2,5 @@
 
 public interface OnNavigationReadyCallback {
 
-  void onNavigationReady();
+  void onNavigationReady(boolean isRunning);
```
Comment (danesfeder) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/OnNavigationReadyCallback.java]: Can we apply the boy-scout rule here and provide some javadoc for this class?  Main explanation will be `isRunning` will be true after you've initialized the `NavigationView` once and began navigation.  Indicates that `NavigationViewModel` is currently running and hasn't been destroyed from a configuration change.

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/camera/NavigationCamera.java [hunk 6]
```diff
@@ -67,6 +67,18 @@ public NavigationCamera(@NonNull MapboxMap mapboxMap, @NonNull MapboxNavigation
     initialize();
   }
 
+  /**
+   * Creates an instance of {@link NavigationCamera}.
+   *
+   * @param mapboxMap for moving the camera
+   * @since 0.15.0
+   */
+  //TODO Could it be package-private?
```
Comment (danesfeder) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/camera/NavigationCamera.java]: I'm fine with either decision here.  If we decide to keep it `public`, let's make sure to also explain in the javadoc that the camera won't do anything until `addProgressChangeListener` is called. 

If we decide against the `public` API, I think we should remove the `camera` package altogether?

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/route/NavigationMapRoute.java [hunk 7]
```diff
@@ -326,6 +336,19 @@ public void showAlternativeRoutes(boolean alternativesVisible) {
...
+  }
+
+  //TODO Is this necessary? It's never used
+  public void removeProgressChangeListener(MapboxNavigation navigation) {
+    if (navigation != null) {
+      navigation.removeProgressChangeListener(progressChangeListener);
+    }
+  }
+
+
```
Comment (danesfeder) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/route/NavigationMapRoute.java]: NIT extra line

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/route/NavigationMapRoute.java [hunk 8]
```diff
@@ -326,6 +336,19 @@ public void showAlternativeRoutes(boolean alternativesVisible) {
     toggleAlternativeVisibility(alternativesVisible);
   }
 
+  public void addProgressChangeListener(MapboxNavigation navigation) {
+    this.navigation = navigation;
+    navigation.addProgressChangeListener(progressChangeListener);
+  }
+
+  //TODO Is this necessary? It's never used
+  public void removeProgressChangeListener(MapboxNavigation navigation) {
```
Comment (danesfeder) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/route/NavigationMapRoute.java]: If we choose to keep add / remove progress change listener `public`, I'd say this is necessary.  We need to keep in mind that public APIs may not be used by us, but could be necessary for developers using these classes as a standalone UI component.

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/camera/NavigationCamera.java (hunk 6 above)
Comment (Guardiola31337) [reply libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/camera/NavigationCamera.java]: Let's leave it `public` for now. We could revisit in a future PR.

**Comment by Guardiola31337** [comment]
@danesfeder addressed all your comments.

This is ready for a second round of 👀

**Comment by Guardiola31337** [comment]
I'll add a new example to the test app so we can test this new functionality with a real `LocationEngine` in a different PR.

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationView.java [hunk 9]
```diff
@@ -253,34 +251,53 @@ public boolean isRecenterButtonVisible() {
...
-    navigationMap.updateWaynameQueryMap(isEnabled);
+    if (navigationMap != null) {
+      navigationMap.updateWaynameQueryMap(isEnabled);
+    }
   }
 
   @Override
   public void takeScreenshot() {
-    map.snapshot(new NavigationSnapshotReadyCallback(this, navigationViewModel));
+    // TODO Refactor -> Move take screenshot functionality into NavigationMapboxMap
```
Comment (danesfeder) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationView.java]: 👍 good call here

**Pull request merged**